Core/Core/Grades/Model/GradeListInteractor.swift

```diff
@@ -310,69 +310,124 @@ public final class GradeListInteractorLive: GradeListInteractor {
+                baseOnGradedAssignments: baseOnGradedAssignments,
+                courseEnrollment: courseEnrollment,
+                gradeEnrollment: gradeEnrollment
+            )
+        }
+    }
+
+    private func getGradeForHideQuantitativeData(
+        baseOnGradedAssignments: Bool,
+        courseEnrollment: Enrollment?,
+        gradeEnrollment: Enrollment?,
+        course: Course
+    ) -> AnyPublisher<String?, Never> {
+        if let gradingPeriodID = gradingPeriodID {
+            return getGradeForGradingPeriod(gradingPeriodID: gradingPeriodID)
+        } else {
+            return getGradeForNoGradingPeriod()
+        }
+
+        func getGradeForGradingPeriod(gradingPeriodID: String) -> AnyPublisher<String?, Never> {
+            let letterGrade = baseOnGradedAssignments
+                ? gradeEnrollment?.currentGrade(gradingPeriodID: gradingPeriodID)
+                : gradeEnrollment?.finalGrade(gradingPeriodID: gradingPeriodID)
+
+            if let letterGrade = letterGrade {
+                return Just(letterGrade).eraseToAnyPublisher()
             } else {
-                if courseEnrollment?.multipleGradingPeriodsEnabled == true, courseEnrollment?.totalsForAllGradingPeriodsOption == false {
-                    return Just(nil).eraseToAnyPublisher()
-                } else if let letterGrade = courseEnrollment?.computedCurrentGrade ?? courseEnrollment?.computedFinalGrade ?? courseEnrollment?.computedCurrentLetterGrade {
-                    return Just(letterGrade).eraseToAnyPublisher()
-                } else {
-                    return Just(courseEnrollment?.convertedLetterGrade(
-                        gradingPeriodID: nil,
-                        gradingScheme: course.gradingScheme
-                    )
-                    ).eraseToAnyPublisher()
-                }
+                return Just(gradeEnrollment?.convertedLetterGrade(
+                    gradingPeriodID: gradingPeriodID,
+                    gradingScheme: course.gradingScheme
+                )).eraseToAnyPublisher()
             }
+        }
+
+        func getGradeForNoGradingPeriod() -> AnyPublisher<String?, Never> {
+            let letterGrade = (
+                baseOnGradedAssignments
+                ? courseEnrollment?.computedCurrentGrade
+                : courseEnrollment?.computedFinalGrade
+            ) ?? courseEnrollment?.computedCurrentLetterGrade
+
+            if courseEnrollment?.multipleGradingPeriodsEnabled == true,
+               courseEnrollment?.totalsForAllGradingPeriodsOption == false {
+                return Just(nil).eraseToAnyPublisher()
+            } else if let letterGrade {
+                return Just(letterGrade).eraseToAnyPublisher()
+            } else {
+                return Just(courseEnrollment?.convertedLetterGrade(
+                    gradingPeriodID: nil,
+                    gradingScheme: course.gradingScheme
+                )).eraseToAnyPublisher()
+            }
```

Comment:
Unless I'm mistaken all the total grade calculation in this class use only `Just()`. I would suggest to simplify this in one of these ways:

1. Use a `var resultLetterGrade: String?` or similar variable, set that value on all the branches and only in the end return it using `Just().erase...`
The same is true for the other places.

2. Or maybe even better to just return optional strings in all these helper methods, and publish them only at the end of the original method.

3. Also,  if you feel like it you could move this to a `TotalGradeProvider` or similar helper class, which could be tested separately. Injecting it to the interactor would allow it to just simply use the returned string.